packages/cli/templates/vue3/src/components/FieldPluginProvider.vue

```diff
@@ -9,7 +10,16 @@ createFieldPlugin((newState) => {
   plugin.type = newState.type
   plugin.error = newState.error
   plugin.data = newState.data
```

Comment:
For another pull request: 

`data.value` is the plain old javascript object that is produced by convertToRaw().

Imagine value has the following shape:

```
{
  "a": 1,
  "b: 2
}
```

And we have two components:

- ComponentA subscribes to `data.value.a`
- ComponentB subscribes to `data.value.b`

Then we mutate "a":

```
{
  "a": 100,
  "b: 2
}
```

Now both ComponentA and ComponentB will re-render, because the reference to `data.value` was not preserved.

In React, we work with immutable data, so this problem does not exist.

In SolidJs, this is done with [Reconcile](https://www.solidjs.com/docs/latest#reconcile):

> Diffs data changes when we can't apply granular updates. Useful for when dealing with immutable data from stores or large API responses.
Should we use the [Immer](https://vuejs.org/guide/extras/reactivity-in-depth.html#immutable-data) library in this case? (edited)

Reply:
good point. I've confirmed the behavior. Now I'm figuring this out.

Reply:
With Immer, we have to provide an updater function, but we always have previous and next data object. So instead of it, I manually assigned each property: https://github.com/storyblok/field-plugin/pull/131/commits/830ea8a0c32f99a0c7bdd97a31052d01da711f76

It's a bit hairy than before, but solves the problem of losing reactive reference. What do you think?

Reply:
That should do part of the trick. This will prevent a change to `data.isModalOpen` from affecting components that only subscribe to `data.value`.

However, without trying it out further, I think a change to `data.value.a` will still re-render `data.value.b`, even if `data.value.b` is a primitive.

Ultimately, I think we may need a change to `@storyblok/field-plugin`, because when the state changes, the library simply [mix in the new state](https://github.com/storyblok/field-plugin/blob/main/packages/field-plugin/src/createFieldPlugin/createPluginActions/createPluginActions.ts#L64) without preserving referential equality for nested properties.

We would need a recursive algorithm, because data.value could have a deep tree structure such as

`data.value`:
```
{
  a: {
    a1: 123,
    a2: 456
  },
  b: [
    {
        b1: 'hello',
    },
    {
       b2: 'world',
    }
  ]
  c: 789,
}
```

`fieldtype-wrapper.js` [does not address this problem](https://github.com/storyblok/storyblok-fieldtype-wrapper/blob/main/src/index.js#L118), so whatever we come up with will be an improvement.

For the library (not Vue), I added [this ticket to the backlog](https://storyblok.atlassian.net/browse/EXT-1409). Shouldn't be a priority.

Reply:
I tested it before pushing this commit, and it actually solves the problem.
In this video, I have a value like `data.value = { count1, count2 }` and updating only one of them causes a render to only the corresponding component, not the other as well.

Of course this doesn't work recursively, but I think it's okay with this level. I don't even think the library should do something about it. It's an intermediary between the field plugin and the Storyblok app, and handles the data via postMessage. So, it's kind of obvious that the object reference is lost in the process. And like we do here, it makes more sense to handle it in each frontend framework level. If it becomes too big, we could have something like `@storyblok/field-plugin/vue`, but I suggest we wait until we get lots of requests for this being handled out of the box.

https://user-images.githubusercontent.com/499898/232793729-8f6f516d-8627-4143-ba72-8a66f496e369.mp4